pywbem_mock/_wbemconnection_mock.py

```diff
@@ -995,6 +1012,104 @@ def _display_objects(obj_type, object_repo, namespace, cmt_begin, cmt_end,
+            DefaultInstanceProvider and it must return data in the same format
+            if the default method returns data.
+        """
+        provider_types = ('instance', 'method')
+
+        if provider_type not in provider_types:
+            raise ValueError("{0!A} is not a valid provider type. "
+                             "Valid provider types are {1!A}.", provider_type,
+                             provider_types)
+
+        if not isinstance(provider, DefaultInstanceProvider):
+            raise ValueError(
+                _format("{0!A} is not a valid subclass of "
+                        "DefaultInstanceProvider. ", provider))
+
+        if classnames is None:
+            raise ValueError(
+                _format('register_provider classnames argument '
+                        'must be string or list '
+                        'of strings. None not allowed.'))
+
+        if namespaces is not None:
+            if not isinstance(namespaces, (list, tuple, six.string_types)):
+                raise ValueError(
+                    _format('register_provider namespaces argument '
+                            '"{0!A}" not list or string'
+                            'of strings.', namespaces))
+
+            if isinstance(namespaces, six.string_types):
+                namespaces = [namespaces]
+
+            for namespace in namespaces:
+                if namespace not in self.namespaces:
+                    raise ValueError(
+                        _format("register_provider namespaces argument "
+                                '"{0!A}" does not exist in {1!A} '
+                                'namespaces.', namespace, namespaces))
+        # If classnames is list, recursively call this method
+        if isinstance(classnames, (list, tuple)):
+            for classname in classnames:
+                self.register_provider(namespaces, classname, provider_type,
+                                       provider)
+            return
+
+        # processing for classnames defining a single classname
+        classname = classnames
+        assert isinstance(classname, six.string_types)
+
+        if namespaces:
+            for namespace in namespaces:
+                if not self.mainprovider.class_exists(namespace, classname):
+                    raise ValueError(
+                        _format("register_provider classname argument "
+                                'class "{0!A}" class does not exist in '
+                                'namespace {1!A}', classname, namespace))
+
+        if classname in self.provider_registry:
+            class_reg = self.provider_registry[classname]
+            if provider_type in class_reg:
+                raise KeyError(
+                    _format("Classname {0!A} cannot register multiple times"
+                            'for provider type {1!A}.',
+                            classname, provider_type))
```

Comment:
I think this restriction should only apply for a combination of namespace and class name. It is valid to register one provider for class A in namespace X and another provider for class A in namespace Y.

Reply:
That one tore up the provider repository structure.  I think I will have to a) drop the any namespace unliess I can find some special flag for it.  b) modify he repository so that it has an additional level for namespace to allow 

TODO

Reply:
DONE. In commit Tuesday night I rewrote register_provider and get_registered_provider. Now we handle multiple providers for single class in multiple namespaces.  As part of that I removed the capability to register for all namespaces.

**DISCUSSION**

Reply:
We agree with this conclusiong.  DONE